Allow clients to send GREASE ESNI extensions.

davidben opened the pull request:
(WG email coming shortly.)

~~NB: When reviewing, ignore the first commit. This PR and https://github.com/tlswg/draft-ietf-tls-esni/pull/124 were originally formulated together. I split them into two based on some feedback, but since they touch the same text, it was simplest to include https://github.com/tlswg/draft-ietf-tls-esni/pull/124 in this PR. If the WG wishes to go with one but not the other, the text and details can be adjusted accordingly.~~

*Update: This has been rebased to master, now that #124 is merged.*

The new server behavior ("Improve ESNI robustness") means clients can send
GREASE ESNI extensions without interop issues. Write some text recommending
this, better meeting the "Do not stick out" criteria.

chris-wood commented:
The GREASE section seems fine to me. (@davidben is there other new material in this PR?) @ekr, @kazuho, please have a look when possible.

davidben commented:
> @davidben is there other new material in this PR?

It's whatever's in the second commit, so mostly that section but also a bit elsewhere:
https://github.com/tlswg/draft-ietf-tls-esni/pull/125/commits/7401f55013f68ec47dad36b99977ad6aa881d9f1

kazuho commented:
I share the view that the GREASE section looks fine.

ekr commented on draft-ietf-tls-esni.md:
```diff
@@ -347,14 +353,41 @@ For servers (in EncryptedExtensions), this extension contains the following
+   enum {
+       esni_required(121),
+   } AlertDescription;
+~~~~
```
What's the function of this alert as opposed to just closing the connection?

ekr commented on draft-ietf-tls-esni.md:
```diff
@@ -446,33 +482,148 @@ to harmonize these to make sure that we maintain key separation.]]
+provide an "encrypted_server_name" extension in EncryptedExtensions, the
+client proceeds with the handshake, verifying the certificate against
+ESNIKeys.public_name as described in {{verify-public-name}}. The client MUST
+NOT enable the False Start optimization {{RFC7918}}. If the handshake completes
```
It's worth noting that this is only relevant for the earlier version case.

ekr commented on draft-ietf-tls-esni.md:
```diff
@@ -446,33 +482,148 @@ to harmonize these to make sure that we maintain key separation.]]
+  certificates, the client MUST abort the connection with an illegal_parameter
+  alert. This case is invalid because {{client-behavior}} requires the client
+  to only offer ESNI-established sessions, and {{server-behavior}} requires
+  the server to decline ESNI-established sessions if it did not accept ESNI.
```
This seems like it ought to be in the client section.

davidben replied:
Well, this is a subsection of the client section. :-) Do you mean that it belongs in {{handle-server-response}}?

The nuisance is it's common to both the key mismatch and rollback cases. The purpose of this section was so the text didn't have to write it twice. (A subroutine if you will.) Though, now that you mention it, it is weird that the subroutine call is labelled "verifying the certificate against ESNIKeys.public_name as described in {{verify-public-name}}" which kinda presupposes a certificate story.

WDYT about this: I've changed the description from "verifying [the certificate] against the public name" to "authenticating for the public name", which puts it at a bit larger scope than certificate verification.

ekr commented on draft-ietf-tls-esni.md:
```diff
@@ -446,33 +482,148 @@ to harmonize these to make sure that we maintain key separation.]]
+  If invalid, it MUST abort the connection with the appropriate alert.
+
+- If the server requests a client certificate, the client MUST respond with an
+  empty Certificate message, denoting no client certificate.
```
IMPORTANT: It seems like with TLS 1.3 you could just abort the connection as soon as you get SFIN.

ekr commented on draft-ietf-tls-esni.md:
```diff
@@ -446,33 +482,148 @@ to harmonize these to make sure that we maintain key separation.]]
+
+- Set the "record_digest" field to a randomly-generated string of hash_length
+  bytes, where hash_length is the length of the hash function associated with
+  the chosen cipher suite.
```
Should we instead use consistent hashing so that you had the same record_digest for each server.

ekr commented on draft-ietf-tls-esni.md:
```diff
@@ -446,33 +482,148 @@ to harmonize these to make sure that we maintain key separation.]]
+
+Offering a GREASE extension is not considered offering an encrypted SNI for
+purposes of requirements in {{client-behavior}}. In particular, the client MAY
+offer sessions established without ESNI.
```
"offer to resume"?

ekr commented on draft-ietf-tls-esni.md:
```diff
@@ -509,8 +660,15 @@ if it were the "server_name" extension to finish the handshake. It
+ignored by servers not negotiating ESNI, including servers which do
+not implement this specification. This may be implemented by adding
+a new field to the server session state which earlier implementations
+cannot parse.
```
I'm not sure how iluminating this hint about how to implement is. I would remove it.

ekr commented on draft-ietf-tls-esni.md:
```diff
@@ -521,57 +679,61 @@ PaddedServerNameList.sni and ClientESNIInner.nonce to the backend
+deployment. This may also occur if a server loses its ESNI keys, or an initial deployment
+of ESNI must be rolled back on the server.
+
+The retry mechanism repairs most such inconsistencies. If server and DNS keys mismatch,
```
I'm not sure if it's "most". We expect CDN conflictst o be common.

davidben commented:
@ekr Did you mean to leave the comments on https://github.com/tlswg/draft-ietf-tls-esni/pull/124? (Notably the text already has seen a number of revisions here. I haven't gone back to update this PR yet. This PR was meant to just be the second commit.)

ekr commented:
Ugh. Having something which claims to be a merged PR is kind of an
attractive nusiance, then, no? Perhaps you can apply the changes here?

On Mon, Feb 25, 2019 at 3:00 PM David Benjamin <notifications@github.com>
wrote:

> @ekr <https://github.com/ekr> Did you mean to leave the comments on #124
> <https://github.com/tlswg/draft-ietf-tls-esni/pull/124>? (Notably the
> text already has seen a number of revisions here. I haven't gone back to
> update this PR yet. This PR was meant to just be the second commit.)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tlswg/draft-ietf-tls-esni/pull/125#issuecomment-467219457>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABD1oTg3ZYafmOt5ubYAqvmNV9qDX1GUks5vRGsRgaJpZM4ZXTha>
> .
>

davidben commented:
In my defense, I split them at your request and did include the following. ;-)

> NB: When reviewing, ignore the first commit. This PR and #124 were originally formulated together. I split them into two based on some feedback, but since they touch the same text, it was simplest to include #124 in this PR. If the WG wishes to go with one but not the other, the text and details can be adjusted accordingly.

But, sure, I can rebase this one. I figured #124 was the bulk of the interesting bits (7401f55013f68ec47dad36b99977ad6aa881d9f1 is much smaller) so I figured we'd resolve #124 first and then I'd update this one at the end.

ekr commented:
On Mon, Feb 25, 2019 at 3:06 PM David Benjamin <notifications@github.com>
wrote:

> In my defense, I split them at your request and did include the following.
> ;-)
>
> NB: When reviewing, ignore the first commit. This PR and #124
> <https://github.com/tlswg/draft-ietf-tls-esni/pull/124> were originally
> formulated together. I split them into two based on some feedback, but
> since they touch the same text, it was simplest to include #124
> <https://github.com/tlswg/draft-ietf-tls-esni/pull/124> in this PR. If
> the WG wishes to go with one but not the other, the text and details can be
> adjusted accordingly.
>
>
Sure, and I read that as "this is an accurate reflection of my put for both
of these".

But, sure, I can rebase this one. I figured #124
> <https://github.com/tlswg/draft-ietf-tls-esni/pull/124> was the bulk of
> the interesting bits (7401f55
> <https://github.com/tlswg/draft-ietf-tls-esni/commit/7401f55013f68ec47dad36b99977ad6aa881d9f1>
> is much smaller) so I figured we'd resolve #124
> <https://github.com/tlswg/draft-ietf-tls-esni/pull/124> first and then
> I'd update this one at the end.
>

Thanks.

-Ekr

—
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tlswg/draft-ietf-tls-esni/pull/125#issuecomment-467221253>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABD1oYi28jBhWjeNTZw4DxJZNo2VxMC_ks5vRGx6gaJpZM4ZXTha>
> .
>

davidben commented:
> Sure, and I read that as "this is an accurate reflection of my put for both of these".

Well, it was when I wrote that sentence. :-) Anyway, it's now been rebased and I've incorporated your comments as well.
